src/Files.App/Filesystem/FilesystemOperations/FilesystemOperations.cs

```diff
@@ -1,996 +1,776 @@
+                    return new StorageHistory(FileOperationType.Copy,
+                        sourceMatch,
+                        await copiedSources.Zip(sourceMatch, (rSrc, oSrc) => new { rSrc, oSrc })
+                            .Select(item => StorageHelpers.FromPathAndType(item.rSrc.Destination, item.oSrc.ItemType)).ToListAsync());
+                }
+                return null; // Cannot undo overwrite operation
             }
-
-            IStorageItem copiedItem = null;
-            //long itemSize = await FilesystemHelpers.GetItemSize(await source.ToStorageItem(associatedInstance));
-
-            if (source.ItemType == FilesystemItemType.Directory)
+            else
             {
-                if (!string.IsNullOrWhiteSpace(source.Path) &&
-                    PathNormalization.GetParentDir(destination).IsSubPathOf(source.Path)) // We check if user tried to copy anything above the source.ItemPath
+                if (copyResult.Items.Any(x => CopyEngineResult.Convert(x.HResult) == FileSystemStatusCode.Unauthorized))
                 {
-                    var destinationName = destination.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries).Last();
-                    var sourceName = source.Path.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries).Last();
-                    ContentDialog dialog = new ContentDialog()
-                    {
-                        Title = "ErrorDialogThisActionCannotBeDone".GetLocalizedResource(),
-                        Content = $"{"ErrorDialogTheDestinationFolder".GetLocalizedResource()} ({destinationName}) {"ErrorDialogIsASubfolder".GetLocalizedResource()} ({sourceName})",
-                        //PrimaryButtonText = "Skip".GetLocalizedResource(),
-                        CloseButtonText = "Cancel".GetLocalizedResource()
-                    };
-
-                    ContentDialogResult result = await this.SetContentDialogRoot(dialog).ShowAsync();
-
-                    if (result == ContentDialogResult.Primary)
+                    if (await RequestAdminOperation())
                     {
-                        progress?.Report(100.0f);
-                        errorCode?.Report(FileSystemStatusCode.InProgress | FileSystemStatusCode.Success);
+                        return await CopyItemsAsync(source, destination, collisions, progress, errorCode, cancellationToken);
                     }
-                    else
+                }
+                else if (copyResult.Items.Any(x => CopyEngineResult.Convert(x.HResult) == FileSystemStatusCode.InUse))
+                {
+                    var failedSources = copyResult.Items.Where(x => CopyEngineResult.Convert(x.HResult) == FileSystemStatusCode.InUse);
+                    var filePath = failedSources.Select(x => x.HResult == CopyEngineResult.COPYENGINE_E_SHARING_VIOLATION_SRC ? x.Source : x.Destination);
```

Comment:
This expression is used several times through the file. Wrapping it up in a method would be worth.